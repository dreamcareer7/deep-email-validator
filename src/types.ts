export {}
declare global {
  interface ObjectConstructor {
    typedKeys<T>(o: T): Array<keyof T>
  }
}
Object.typedKeys = Object.keys as any
