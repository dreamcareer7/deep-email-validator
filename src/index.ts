import isEmail from 'validator/lib/isEmail'
import { OutputFormat, createOutput } from './output/output'
import { checkTypo } from './typo/typo'
import { getBestMx } from './dns/dns'
import { checkSMTP } from './smtp/smtp'
import { checkDisposable } from './disposable/disposable'

declare global {
  interface ObjectConstructor {
    typedKeys<T>(o: T): Array<keyof T>
  }
}
Object.typedKeys = Object.keys as any

export async function validate(
  recipient: string,
  sender: string = 'name@example.org'
): Promise<OutputFormat> {
  if (!isEmail(recipient)) return createOutput('regex')

  const typoResponse = await checkTypo(recipient)
  if (typoResponse) return createOutput('typo', typoResponse)

  const domain = recipient.split('@')[1]

  const disposableResponse = await checkDisposable(domain)
  if (disposableResponse) return createOutput('disposable', disposableResponse)

  const mx = await getBestMx(domain)
  if (!mx) return createOutput('mx', 'MX record not found')

  return checkSMTP(sender, recipient, mx.exchange)
}

async function main() {
  if (process.argv[2]) {
    const res = await validate(process.argv[2])
    console.log(JSON.stringify(res, null, 2))
  }
}
main()
