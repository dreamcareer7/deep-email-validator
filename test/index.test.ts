import { validate } from '../src/index'
import _ from 'lodash'

describe('validation tests', () => {
  it('fails with bad regex', async () => {
    const res = await validate('dav id@gmail.com')
    expect(res).toMatchSnapshot()
  })
  it('fails with common typo', async () => {
    const res = await validate('david@gmaill.com')
    expect(res).toMatchSnapshot()
  })
  // it('fails with bad dns', async () => {
  //   const res = await validate('david@capoeira.london')
  //   expect(res).toMatchSnapshot()
  // })
  it('fails with bad username', async () => {
    const res = await validate('david@andco.life')
    expect(res).toMatchSnapshot()
  })
})
