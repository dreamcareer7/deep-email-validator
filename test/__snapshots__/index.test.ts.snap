// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validation tests bad username 1`] = `
Object {
  "valid": false,
  "validators": Object {
    "disposable": Object {
      "reason": undefined,
      "valid": false,
    },
    "mx": Object {
      "reason": undefined,
      "valid": false,
    },
    "regex": Object {
      "reason": undefined,
      "valid": false,
    },
    "smtp": Object {
      "reason": "Invalid Mailbox",
      "valid": false,
    },
    "typo": Object {
      "reason": undefined,
      "valid": false,
    },
  },
}
`;

exports[`validation tests fails with bad dns 1`] = `
Object {
  "valid": false,
  "validators": Object {
    "disposable": Object {
      "reason": undefined,
      "valid": false,
    },
    "mx": Object {
      "reason": "MX record not found",
      "valid": false,
    },
    "regex": Object {
      "reason": undefined,
      "valid": false,
    },
    "smtp": Object {
      "reason": undefined,
      "valid": false,
    },
    "typo": Object {
      "reason": undefined,
      "valid": false,
    },
  },
}
`;

exports[`validation tests fails with bad regex 1`] = `
Object {
  "valid": false,
  "validators": Object {
    "disposable": Object {
      "reason": undefined,
      "valid": false,
    },
    "mx": Object {
      "reason": undefined,
      "valid": false,
    },
    "regex": Object {
      "reason": "Invalid regex",
      "valid": false,
    },
    "smtp": Object {
      "reason": undefined,
      "valid": false,
    },
    "typo": Object {
      "reason": undefined,
      "valid": false,
    },
  },
}
`;

exports[`validation tests fails with bad username 1`] = `
Object {
  "valid": false,
  "validators": Object {
    "disposable": Object {
      "reason": undefined,
      "valid": false,
    },
    "mx": Object {
      "reason": undefined,
      "valid": false,
    },
    "regex": Object {
      "reason": undefined,
      "valid": false,
    },
    "smtp": Object {
      "reason": "Invalid Mailbox",
      "valid": false,
    },
    "typo": Object {
      "reason": undefined,
      "valid": false,
    },
  },
}
`;

exports[`validation tests fails with common typo 1`] = `
Object {
  "valid": false,
  "validators": Object {
    "disposable": Object {
      "reason": undefined,
      "valid": false,
    },
    "mx": Object {
      "reason": undefined,
      "valid": false,
    },
    "regex": Object {
      "reason": undefined,
      "valid": false,
    },
    "smtp": Object {
      "reason": undefined,
      "valid": false,
    },
    "typo": Object {
      "reason": "Likely typo, suggested email: david@gmail.com",
      "valid": false,
    },
  },
}
`;
